src/treesitter.rs

```diff
@@ -21,3 +30,305 @@ impl From<&'_ Node<'_>> for Match {
+                RopeOrSliceTextProviderIterator::Rope(RopeOrSliceRopeTextProviderIterator::new(
+                    rope_slice,
+                    |rope_slice| rope_slice.chunks(),
+                ))
+            }
+        }
+    }
+}
+
+impl<'a> TextProvider<'a> for &'a RopeOrSlice<'a> {
+    type I = RopeOrSliceTextProviderIterator<'a>;
+
+    fn text(&mut self, node: Node) -> Self::I {
+        match self {
+            RopeOrSlice::Slice(slice) => {
+                RopeOrSliceTextProviderIterator::Slice(iter::once(&slice[node.byte_range()]))
+            }
+            RopeOrSlice::Rope(rope) => {
+                let rope_slice = rope.byte_slice(node.byte_range());
+                RopeOrSliceTextProviderIterator::Rope(RopeOrSliceRopeTextProviderIterator::new(
+                    rope_slice,
+                    |rope_slice| rope_slice.chunks(),
+                ))
+            }
+        }
+    }
+}
+
+impl<'a> Parseable for RopeOrSlice<'a> {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        match self {
+            Self::Slice(slice) => slice.parse(parser, old_tree),
+            Self::Rope(rope) => rope.parse(parser, old_tree),
+        }
+    }
+}
+
+impl<'a> Parseable for &'a RopeOrSlice<'a> {
+    fn parse(&self, parser: &mut Parser, old_tree: Option<&Tree>) -> Option<Tree> {
+        match self {
+            RopeOrSlice::Slice(slice) => slice.parse(parser, old_tree),
+            RopeOrSlice::Rope(rope) => rope.parse(parser, old_tree),
+        }
+    }
+}
+
+impl<'a> From<&'a [u8]> for RopeOrSlice<'a> {
+    fn from(value: &'a [u8]) -> Self {
+        Self::Slice(value)
+    }
+}
+
+impl<'a> From<&'a Rope> for RopeOrSlice<'a> {
+    fn from(value: &'a Rope) -> Self {
+        Self::Rope(value)
+    }
+}
+
+pub enum RopeOrSliceTextProviderIterator<'a> {
+    Slice(iter::Once<&'a [u8]>),
+    Rope(RopeOrSliceRopeTextProviderIterator<'a>),
+}
+
+impl<'a> Iterator for RopeOrSliceTextProviderIterator<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::Slice(slice_iterator) => slice_iterator.next(),
+            Self::Rope(rope_iterator) => rope_iterator.next().map(str::as_bytes),
+        }
+    }
+}
+
+#[self_referencing]
+pub struct RopeOrSliceRopeTextProviderIterator<'a> {
+    rope_slice: RopeSlice<'a>,
+
+    #[borrows(rope_slice)]
+    chunks_iterator: Chunks<'a>,
+}
+
+impl<'a> Iterator for RopeOrSliceRopeTextProviderIterator<'a> {
+    type Item = &'a str;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.with_chunks_iterator_mut(|chunks_iterator| chunks_iterator.next())
+    }
+}
+
+// I believe this type can't be Copy/Clone in order for the
+// `get_captures()` unsafe stuff to be sound
+pub struct CaptureInfo<'a> {
+    pub node: Node<'a>,
+    pub pattern_index: usize,
+}
+
+#[self_referencing]
+pub struct Captures<'a, 'text: 'a, 'tree: 'a> {
```

Comment:
Ok so for the `get_captures()` iterator type again we're in self-referencing struct world

Why? Because the items returned by a `tree_sitter::QueryCaptures` are lifetime-dependent on things like the tree, text and query that passed to `QueryCursor::captures()` (and I think that implies that?)

And the query cursor itself wants to be created inside the shared abstraction but the captures iterator is also dependent on its lifetime I believe

So basically "toss all the stuff that the iterator depends on as fields and then use `ouroboros` to make the one iterator field self-referentially dependent on those"

But then it didn't want to implement `Iterator`, I think basically because items returned by the iterator would be lifetime-dependent on the struct (aka `Iterator`) itself and `Iterator` does not support that (your returned items cannot depend on the lifetime of the `Iterator` itself)

So then looked like `streaming_iterator` was a reasonable option for trying to implement a "lending" iterator on this struct

I was able to get that to basically work but see comments below re a little `unsafe`